crossbeam-utils/src/cyclic_timestamp.rs

```diff
@@ -0,0 +1,529 @@
+        if self.0 != 0xbfffffffffffffff {
+            CyclicTimestamp(self.0 + 1)
+        } else {
+            CyclicTimestamp(0)
+        }
+    }
+}
+
+impl Timestamp for CyclicTimestamp<u128> {
+    fn new() -> CyclicTimestamp<u128> {
+        CyclicTimestamp(0)
+    }
+
+    fn earlier(&self, other: &CyclicTimestamp<u128>) -> bool {
+        let self_number = 0x3fffffffffffffffffffffffffffffff & self.0;
+        let other_number = 0x3fffffffffffffffffffffffffffffff & other.0;
+        let self_rps = self.0 & 0xc0000000000000000000000000000000;
+        let other_rps = other.0 & 0xc0000000000000000000000000000000;
+
+        (self_rps == other_rps && self_number < other_number) ||
+            ((self_rps == 0x00000000000000000000000000000000) &&
+             (other_rps == 0x40000000000000000000000000000000)) ||
+            ((self_rps == 0x40000000000000000000000000000000) &&
+             (other_rps == 0x80000000000000000000000000000000)) ||
+            ((self_rps == 0x80000000000000000000000000000000) &&
+             (other_rps == 0x00000000000000000000000000000000))
+    }
+
+    fn later(&self, other: &CyclicTimestamp<u128>) -> bool {
+        let self_number = 0x3fffffffffffffffffffffffffffffff & self.0;
+        let other_number = 0x3fffffffffffffffffffffffffffffff & other.0;
+        let self_rps = self.0 & 0xc0000000000000000000000000000000;
+        let other_rps = other.0 & 0xc0000000000000000000000000000000;
+
+        (self_rps == other_rps && self_number > other_number) ||
+            ((self_rps == 0x40000000000000000000000000000000) &&
+             (other_rps == 0x00000000000000000000000000000000)) ||
+            ((self_rps == 0x80000000000000000000000000000000) &&
+             (other_rps == 0x40000000000000000000000000000000)) ||
+            ((self_rps == 0x00000000000000000000000000000000) &&
+             (other_rps == 0x80000000000000000000000000000000))
+    }
+
+    fn succ(&self) -> CyclicTimestamp<u128> {
+        if self.0 != 0xbfffffffffffffffffffffffffffffff {
+            CyclicTimestamp(self.0 + 1)
+        } else {
+            CyclicTimestamp(0)
+        }
+    }
+}
+
+#[test]
+fn test_u8_wrap() {
+    assert_eq!(CyclicTimestamp(0 as u8), CyclicTimestamp(0xbf as u8).succ())
+}
+
+#[test]
+fn test_u16_wrap() {
+    assert_eq!(CyclicTimestamp(0 as u16),
+               CyclicTimestamp(0xbfff as u16).succ())
+}
+
+#[test]
+fn test_u32_wrap() {
+    assert_eq!(CyclicTimestamp(0 as u32),
+               CyclicTimestamp(0xbfffffff as u32).succ())
+}
+
+#[test]
+fn test_u64_wrap() {
+    assert_eq!(CyclicTimestamp(0 as u64),
+               CyclicTimestamp(0xbfffffffffffffff as u64).succ())
+}
+
+#[test]
+fn test_u128_wrap() {
+    assert_eq!(CyclicTimestamp(0 as u128),
+               CyclicTimestamp(0xbfffffffffffffffffffffffffffffff as u128)
+                 .succ())
+}
+
+#[test]
+fn test_u8() {
+    let values: [CyclicTimestamp<u8>; 9] = [
+        CyclicTimestamp(0x00),
```

Comment:
For this and other examples, it would be good to give some explanation of why these specific bit patterns are chosen (or that they're simply random) and maybe give them names if they're significant.